[Merged by Bors] - feat(analysis/normed_space/basic): add `normed_division_ring`

j-loreaux opened the pull request:
This defines normed division rings and generalizes some of the lemmas that applied to normed fields instead to normed division rings.

This breaks one `ac_refl`; although this could be resolved by using `simp only {canonize_instances := tt}` first, it's easier to just tweak the proof.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
-->
- [x] depends on: #12163

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

j-loreaux committed:
add normed division ring

eric-wieser commented on src/analysis/normed_space/basic.lean:
```diff
@@ -80,6 +92,10 @@ instance semi_normed_comm_ring.to_comm_ring [β : semi_normed_comm_ring α] : co
 instance normed_ring.to_normed_group [β : normed_ring α] : normed_group α := { ..β }
 
+@[priority 100] -- see Note [lower instance priority]
+instance normed_division_algebra.to_normed_group [β : normed_division_ring α] : normed_group α :=
```
I don't think this instance should exist - instead you should have
```suggestion
instance normed_division_ring.to_normed_ring [β : normed_division_ring α] : normed_ring α :=
```
(And then the instance you had originally will be found transitively)

j-loreaux replied:
I already had the `to_normed_ring` one, I'm not sure why I add this. Removed.

j-loreaux committed:
remove the normed_group instance

j-loreaux committed:
delete normed_field lemmas

This does several things:
1. Moves the division ring lemmas into root namespace
2. Deletes the corresponding lemmas from the normed field namespace
3. Deletes the `normed_field.` references across mathlib

j-loreaux committed:
add list.norm_prod and list.nnnorm_prod

j-loreaux committed:
rename instances

j-loreaux committed:
fix a stray deletion of `normed_field.` which was needed.

j-loreaux committed:
remove extraneous `normed_field.` deletions

j-loreaux committed:
merge master

j-loreaux committed:
fix `analysis.normed_space.basic`

j-loreaux committed:
remove stray `x`

eric-wieser commented on src/analysis/normed_space/basic.lean:
```diff
@@ -282,6 +279,16 @@ export norm_one_class (norm_one)
+lemma list.nnnorm_prod : ∀ l : list α, ∥l.prod∥₊ = (l.map nnnorm).prod
+| [] := by simp
+| (a :: l) :=
+  by rw [list.map_cons, list.prod_cons, @list.prod_cons _ _ _ ∥a∥₊, nnnorm_mul, list.nnnorm_prod]
```
Can you prove this with `(norm_hom.to_monoid_hom : α →* ℝ).map_list_prod _` or similar?

j-loreaux replied:
ah, nice. Yes, that works.

eric-wieser commented on src/analysis/normed_space/basic.lean:
```diff
@@ -316,6 +315,39 @@ nnreal.eq $ by simp
+(nnnorm_hom.to_monoid_hom : α →* ℝ≥0).map_prod f s
+
 @[priority 100] -- see Note [lower instance priority]
 instance normed_field.has_continuous_inv₀ : has_continuous_inv₀ α :=
```
I'm pretty sure this is easy to generalize to `division_ring`, you just have to use `∥r∥⁻¹ * ∥r - e∥ * ∥e∥⁻¹` or similar below instead of `∥r - e∥ / ∥r∥ / ∥e∥`

j-loreaux replied:
yes, done.

j-loreaux committed:
simplify proof of `list.norm_prod` with `norm_hom` (and `nnnorm`)

eric-wieser commented on src/analysis/normed_space/basic.lean:
```diff
@@ -288,20 +285,18 @@ nnreal.eq $ norm_mul a b
+lemma list.norm_prod (l : list α) : ∥l.prod∥ = (l.map norm).prod :=
+(norm_hom.to_monoid_hom : α →* ℝ).map_list_prod _
+
+lemma list.nnnorm_prod (l : list α) : ∥l.prod∥₊ = (l.map nnnorm).prod :=
```
For the sake of minimizing the diff, can you leave these lemmas in the same place as the old prod lemmas?

j-loreaux committed:
generalize `has_continuous_inv0` to normed division ring

j-loreaux committed:
workaround ac_refl

j-loreaux committed:
Merge branch 'master' into j-loreaux/normed-division-ring

eric-wieser commented on src/analysis/normed_space/basic.lean:
```diff
@@ -299,13 +296,11 @@ nnreal.eq $ norm_mul a b
-  ∥∏ b in s, f b∥ = ∏ b in s, ∥f b∥ :=
-(norm_hom.to_monoid_hom : α →* ℝ).map_prod f s
+lemma list.norm_prod (l : list α) : ∥l.prod∥ = (l.map norm).prod :=
+(norm_hom.to_monoid_hom : α →* ℝ).map_list_prod _
```
For another reviewer: is this name better or worse than `norm_list_prod`?

YaelDillies replied:
I prefer `norm_list_prod` to match `map_list_prod` and friends.

j-loreaux replied:
But doesn't `list.norm_prod` give you dot notation?

YaelDillies replied:
Eh yeah, it does. Both are fine, really. If you want to keep it like that, add `protected` at least.

eric-wieser reviewed (approved):
Will let another reviewer finish up, this looks good now.

It might be worth having a quick skim for any lemmas that can be relaxed from `normed_field` to `normed_division_ring` in other files.

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into j-loreaux/normed-division-ring

Conflicts resolved with `git diff $(git merge-base HEAD^ origin/master)..HEAD^ -- src/analysis/normed_space/basic.lean | patch src/analysis/normed/normed_field.lean`

YaelDillies commented on src/analysis/normed/normed_field.lean:
```diff
@@ -57,6 +57,18 @@ class normed_ring (α : Type*) extends has_norm α, ring α, metric_space α :=
 (dist_eq : ∀ x y, dist x y = norm (x - y))
 (norm_mul : ∀ a b, norm (a * b) ≤ norm a * norm b)
 
+/-- A normed division ring is a division ring endowed with a seminorm which satisfies the inequality
```
```suggestion
/-- A normed division ring is a division ring endowed with a seminorm which satisfies the equality
```

YaelDillies commented on src/analysis/normed/normed_field.lean:
```diff
@@ -375,22 +370,58 @@ nnreal.eq $ by simp
 
 @[priority 100] -- see Note [lower instance priority]
-instance normed_field.has_continuous_inv₀ : has_continuous_inv₀ α :=
+instance normed_division_ring.has_continuous_inv₀ : has_continuous_inv₀ α :=
```
```suggestion
instance normed_division_ring.to_has_continuous_inv₀ : has_continuous_inv₀ α :=
```

YaelDillies commented on src/analysis/normed/normed_field.lean:
```diff
@@ -375,22 +370,58 @@ nnreal.eq $ by simp
+
+@[priority 100] -- see Note [lower instance priority]
+instance normed_field.to_normed_division_ring : normed_division_ring α :=
+{ norm_mul' := normed_field.norm_mul', ..‹normed_field α› }
```
```suggestion
{ ..‹normed_field α› }
```

YaelDillies commented on src/analysis/normed/normed_field.lean:
```diff
@@ -57,6 +57,18 @@ class normed_ring (α : Type*) extends has_norm α, ring α, metric_space α :=
+/-- A normed division ring is a division ring endowed with a seminorm which satisfies the inequality
+`∥x y∥ = ∥x∥ ∥y∥`. -/
+class normed_division_ring (α : Type*) extends has_norm α, division_ring α, metric_space α :=
+(dist_eq : ∀ x y, dist x y = norm (x - y))
+(norm_mul' : ∀ a b, norm (a * b) = norm a * norm b)
```
Very minor, but pick a side in the letters war.

j-loreaux replied:
This is like this throughout the file, who knows why. I didn't change it here or elsewhere only to minimize the diff.

YaelDillies commented on src/analysis/normed/normed_field.lean:
```diff
@@ -375,22 +370,58 @@ nnreal.eq $ by simp
+`0` and `1`. This makes it possible to bring any element arbitrarily close to `0` by multiplication
+by the powers of any element, and thus to relate algebra and topology. -/
+class nondiscrete_normed_field (α : Type*) extends normed_field α :=
+(non_trivial : ∃x:α, 1<∥x∥)
```
```suggestion
(non_trivial : ∃ x : α, 1 < ∥x∥)
```
Btw, this field name is weird.

YaelDillies reviewed (commented):
Minor things

j-loreaux committed:
apply suggestions from code review

j-loreaux committed:
protect list.norm_prod

j-loreaux committed:
Merge branch 'master' into j-loreaux/normed-division-ring

eric-wieser commented on src/analysis/asymptotics/asymptotics.lean:
```diff
@@ -977,7 +977,7 @@ begin
   convert mul_le_mul hx₁ hx₂ (norm_nonneg _) (le_trans (norm_nonneg _) hx₁) using 1,
   rw norm_mul,
-  ac_refl
+  ring,
```
`rw mul_mul_mul_comm` works here too, and you can merge it with the line above.

j-loreaux committed:
make quaternions a normed_division_ring

j-loreaux committed:
use mul_mul_mul_comm

eric-wieser reviewed (approved):
Thanks!

bors d+

Let's hope CI doesn't fail mysteriously this time
